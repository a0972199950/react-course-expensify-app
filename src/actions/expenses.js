import database from "../firebase/firebase";

export const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense    
});

export const startAddExpense = (expenseData = {}) => {
    // 藉由在store裡面傳入redux-thunk，我們可以讓action不要return一個object，而是return一個函數進去store
    // redux-thunk會把dispatch傳進我們return的函數作為第一個參數來讓我們使用
    return (dispatch) => {
        // 解構expenseData
        const {description = "", note = "", amount = 0, createdAt = 0} = expenseData;
        const expense = { description, note, amount, createdAt };
        // push資料進firebase 
        // 這邊的return會return一個Promise，是為了給tests用的
        // 之後會在tests裡面再對上面的Promise連接.then()
        return database
            .ref("expenses")
            .push(expense)
            .then((ref) => {
                dispatch(addExpense({
                    id: ref.key,
                    ...expense
                }))
            })
    }
}


export const removeExpense = (id = undefined) => ({
    type: "REMOVE_EXPENSE",
    id
});


export const startRemoveExpense = (id = undefined) => {
    return (dispatch) => {
        return database
            .ref("expenses/" + id)
            .remove()
            .then(() => {
                dispatch(removeExpense(id));
            })
            .catch((e) => {
                console.log("remove failed", e);
            });

    };
};


export const editExpense = (id, updates) => ({
    type: "EDIT_EXPENSE",
    id,
    updates
});


export const startEditExpense = (id, updates) => {
    return (dispatch) => {
        return database
            .ref("expenses/" + id)
            .update(updates)
            .then(() => {
                dispatch(editExpense(id, updates));
            })
            .catch((e) => {
                console.log("update failed", e);
            });
    };
};


// 將從database拿到的資料寫到redux裡面
export const setExpenses = (expenses) => {
    return {
        type: "SET_EXPENSES",
        expenses
    }
};


// 從database獲取既存的資料
export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref("expenses").once("value").then((snapshot) => {
            const expensesArray = [];

            snapshot.forEach((childSnapshot) => {
                expensesArray.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setExpenses(expensesArray));

        }).catch((e) => {
            console.log("fatch data from database failed", e);
        });
    };
};