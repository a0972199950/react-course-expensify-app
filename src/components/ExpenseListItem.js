import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";

const ExpenseListItem = ({description, amount, createdAt, id}) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>${amount} + time:{moment(createdAt).format("YY/MM/DD hh:mm:ss")}</p>
    </div>
);



export default ExpenseListItem;