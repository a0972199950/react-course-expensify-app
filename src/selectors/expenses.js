import moment from "moment";

export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {

        const startDateMatch = !startDate || moment(expense.createdAt).isSameOrAfter(startDate, "date");
        const endDateMatch = !endDate || moment(expense.createdAt).isSameOrBefore(endDate, "date");
        const textMatch = !text || expense.description.includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((last, next) => {
        switch(sortBy){
            case "amount":
                return (last.amount < next.amount) ? -1 : 1;

            case "date": 
                return (last.createdAt < next.createdAt) ? -1 : 1;

            default:
                return 0;
        };
    });

};