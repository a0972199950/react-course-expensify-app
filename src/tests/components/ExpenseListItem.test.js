import React from "react";
import ExpenseListItem from "../../components/ExpenseListItem";
import { shallow } from "enzyme";
import expenses from "../fixtures/expense";

expenses.forEach((expense) => {
    test("should render ExpenseListItem with expense", () => {
        const wrapper = shallow(<ExpenseListItem {...expense}/>);
    
        expect(wrapper).toMatchSnapshot();
    })
})
