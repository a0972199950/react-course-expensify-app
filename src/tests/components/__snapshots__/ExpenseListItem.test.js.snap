// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with expense 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      rent
    </h3>
  </Link>
  <p>
    $
    1000
     + time:
    70/01/01 08:00:00
  </p>
</div>
`;

exports[`should render ExpenseListItem with expense 2`] = `
<div>
  <Link
    replace={false}
    to="/edit/2"
  >
    <h3>
      coffee
    </h3>
  </Link>
  <p>
    $
    5000
     + time:
    69/12/29 08:00:00
  </p>
</div>
`;

exports[`should render ExpenseListItem with expense 3`] = `
<div>
  <Link
    replace={false}
    to="/edit/3"
  >
    <h3>
      dinner
    </h3>
  </Link>
  <p>
    $
    3000
     + time:
    70/01/04 08:00:00
  </p>
</div>
`;
